mirror_node_contract_estimate_gas_query.go

```diff
@@ -0,0 +1,74 @@
+
+// SPDX-License-Identifier: Apache-2.0
+
+// MirrorNodeContractCallQuery returns a result from EVM gas estimation of read-write operations.
+type MirrorNodeContractEstimateGasQuery struct {
+	mirrorNodeContractQuery
+}
+
+func NewMirrorNodeContractEstimateGasQuery() *MirrorNodeContractEstimateGasQuery {
+	query := new(MirrorNodeContractEstimateGasQuery)
+	return query
+}
+
+// SetContractID sets the contract instance to call.
+func (query *MirrorNodeContractEstimateGasQuery) SetContractID(contractID ContractID) *MirrorNodeContractEstimateGasQuery {
+	query.setContractID(contractID)
+	return query
+}
+
+// SetContractEvmAddress sets the 20-byte EVM address of the contract to call.
+func (query *MirrorNodeContractEstimateGasQuery) SetContractEvmAddress(contractEvmAddress string) *MirrorNodeContractEstimateGasQuery {
+	query.setContractEvmAddress(contractEvmAddress)
+	return query
+}
+
+// SetSender sets the sender of the transaction simulation.
+func (query *MirrorNodeContractEstimateGasQuery) SetSender(sender AccountID) *MirrorNodeContractEstimateGasQuery {
+	query.setSender(sender)
+	return query
+}
+
+// SetSenderEvmAddress sets the 20-byte EVM address of the sender of the transaction simulation.
+func (query *MirrorNodeContractEstimateGasQuery) SetSenderEvmAddress(senderEvmAddress string) *MirrorNodeContractEstimateGasQuery {
+	query.setSenderEvmAddress(senderEvmAddress)
+	return query
+}
+
+// SetFunction sets the function parameters as their raw bytes.
+func (query *MirrorNodeContractEstimateGasQuery) SetFunction(name string, params *ContractFunctionParameters) *MirrorNodeContractEstimateGasQuery {
+	query.setFunction(name, params)
+	return query
+}
+
+// SetFunction sets the function parameters as their raw bytes.
+func (query *MirrorNodeContractEstimateGasQuery) SetFunctionParameters(byteArray []byte) *MirrorNodeContractEstimateGasQuery {
+	query.setFunctionParameters(byteArray)
+	return query
+}
+
+// SetValue sets the amount of value (in tinybars or wei) to be sent to the contract in the transaction.
+func (query *MirrorNodeContractEstimateGasQuery) SetValue(value int64) *MirrorNodeContractEstimateGasQuery {
+	query.setValue(value)
+	return query
+}
+
+// SetGasLimit sets the gas limit for the contract call.
+// This specifies the maximum amount of gas that the transaction can consume.
+func (query *MirrorNodeContractEstimateGasQuery) SetGasLimit(gasLimit int64) *MirrorNodeContractEstimateGasQuery {
+	query.setGasLimit(gasLimit)
+	return query
+}
+
+// SetGasPrice sets the gas price to be used for the contract call.
+// This specifies the price of each unit of gas used in the transaction.
+func (query *MirrorNodeContractEstimateGasQuery) SetGasPrice(gasPrice int64) *MirrorNodeContractEstimateGasQuery {
+	query.setGasPrice(gasPrice)
+	return query
+}
+
+// Returns gas estimation for the EVM execution
+func (query *MirrorNodeContractEstimateGasQuery) Execute(client *Client) (uint64, error) {
+	return query.estimateGas(client)
+}
```

Comment:
Is there any difference between `MirrorNodeContractEstimateGasQuery` and `MirrorNodeContractCallQuery` if not for this `execute` method. You are housing the logic for `estimateGas` and the `call` in the parent which is `query`. Going down this path you would have to extend the logic in `query` while creating files and structs which just hold all the properties of the `query` and just adding an unnecessary layer of setters and getters. I would see the need of having `MirrorNodeContractEstimateGasQuery` and `MirrorNodeContractCallQuery` if they held at least some specific fields or functionalities.

Reply:
We are doing it solely for the naming. I initially had them be 1 query, which had `estimate` and `call` methods, but Simi and Nana wanted the query name to say what the query does - hence how we have 2 queries - `EstimateGasQuery` and `CallQuery`.